### PR TITLE
Add write option for geopackage and geoparquet

### PR DESCRIPTION
**Pull request by val-ismaili**
I've add the option to write to geopackage or geoparquet. default is changed from `.geojson` to `.gpkg`

the new cli command looks like:
`osmox run configs/example.json example/isle-of-man-latest.osm.pbf isle-of-man -f geopackage  -crs epsg:27700 -l`

To allow write to geoparquet I had to `pip install pyarrow`. What's the best way of updating the environment for the repo?

### CODE REVIEW

File: src/osmox/cli.py [hunk 1]
```diff
@@ -107,17 +114,31 @@ def run(config_path, input_path, output_name, crs, single_use, lazy):
...
+    output_filename = f"{output_name}_{crs.replace(':', '_')}.{extension}"
+
+    if format != "geoparquet":
+        writer(output_filename, driver=driver)
+    else:
+        writer(output_filename)
+
+    if not crs == "epsg:4326" and format != "geoparquet":
+        logger.info(" Reprojecting output to EPSG:4326 (lat lon)")
+        gdf.to_crs("epsg:4326").to_file(f"{output_name}_epsg_4326.{extension}", driver=driver)
```
Comment (brynpickering) [inline src/osmox/cli.py:139]: Based on earlier comments, this becomes:

```python
if pyproj.CRS(crs) != pyproj.CRS("epsg:4326"):
    gdf_4326 = gdf.to_crs("epsg:4326")
    getattr(gdf_4326, writer)( f"{output_name}_epsg_4326.{extension}", **kwargs)
```

File: src/osmox/cli.py [hunk 2]
```diff
@@ -107,17 +114,31 @@ def run(config_path, input_path, output_name, crs, single_use, lazy):
...
+        extension = "geojson"
+        driver = 'GeoJSON'
+        writer = gdf.to_file
+    elif format == "geopackage":
+        extension = "gpkg"
+        driver = 'GPKG'
+        writer = gdf.to_file
+    elif format == "geoparquet":
+        extension = "parquet"
+        writer = gdf.to_parquet
```
Comment (brynpickering) [inline src/osmox/cli.py:127]: `writer = "to_parquet"` etc. and then later `getattr(gdf, writer)(...)`, this means you don't have to have the extra conditional when you reproject and save later.

**Review by brynpickering — CHANGES_REQUESTED** [review]
Looks good. A range of suggestions for making things a bit slimmer and possibly more maintanable (fewer places to make changes in future). 

Also, this would be a good opportunity to add a CLI test. You can check out how it's done in pam: https://github.com/arup-group/pam/blob/main/tests/test_99_cli.py (you basically create a dummy CLI call and then check that it did what you expected it to).

Note that when storing data, the tests do so to a temporary path `tmp_path`. This is a pytest fixture: https://docs.pytest.org/en/6.2.x/tmpdir.html

**Comment by brynpickering** [comment]
RE pyarrow: add it to `requirements/base.txt`. This is the file containing all dependencies that the module depends upon. You should set some loose pinning according to the latest pyarrow release (`pyarrow >= 15, < 16`?)

**Comment by val-ismaili** [comment]
@brynpickering I've refactored `cli.py` as your suggestions (I think!). Taken a stab at testing cli. Figured the two things to test would be:
a) by excluding optional arguments, it uses the correct defaults, and;
b) each of the three output format types work

The tests currently both fail...and I'm not sure what's going on...

File: tests/test_cli.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        ]
+    )
+    #print(result.output)
+    assert result.exit_code == 0
+    assert "geopackage" in result.exit_code
+    assert "epsg:4326" in result.exit_code
+
+
+def test_cli_output_formats(runner, config_path, toy_osm_path):
+    for output_format in ["geojson", "geopackage", "geoparquet"]:
```
Comment (brynpickering) [inline tests/test_cli.py:49]: In tests, for loops are moved to a wrapper around the method: [pytest.mark.parameterize](https://docs.pytest.org/en/7.1.x/how-to/parametrize.html).

Here:

```python
@pytest.mark.parametrize("output_format", ["geojson", "geopackage", "geoparquet"])
def test_cli_output_formats(runner, config_path, toy_osm_path, output_format):
   result = runner.invoke(...
```

**Review by brynpickering — CHANGES_REQUESTED** [review]
Minor comments.

Understandable that you're having trouble with the CLI test debugging - it's always a bit of a pain. I've pushed a change with a helper function to add the full error traceback. You can hopefully use that traceback to debug the issue (something to do with the progress bar helper function).

Another point: don't forget to install pre-commit in your environment to have the code formatted correctly on commit (`pre-commit install`).

**Comment by brynpickering** [comment]
Great @val-ismaili ! Now just need to run `pre-commit install` and `pre-commit run --all` in your local repo so that it formats all the files to make the `pre-commit.ci` bot happy. This will make sure that files you previously committed without `pre-commit` installed will be retrospectively formatted

**Comment by val-ismaili** [comment]
@brynpickering yeah i was running pre-commit but still getting the issue. unsure what im missing...

```
(osmox) val.ismaili@LDNML6TDXNNY osmox % pre-commit run --all
check for added large files..............................................Passed
black....................................................................Passed
ruff.....................................................................Failed
- hook id: ruff
- exit code: 1
- files were modified by this hook

Found 4 errors (4 fixed, 0 remaining).
```

**Comment by brynpickering** [comment]
Once you run it, you should have 4 files with changes that you can then commit - right?

**Comment by val-ismaili** [comment]
i did yes, which is what the previous commit was here (https://github.com/arup-group/osmox/pull/42/commits/8830df3f1dd69734c15cdc53637fa916f35e18c3). now when i run `pre-commit run --all` it shows all as passing which is why im confused why its failing on the PR.

```
(osmox) val.ismaili@LDNML6TDXNNY osmox % pre-commit run --all
check for added large files..............................................Passed
black....................................................................Passed
ruff.....................................................................Passed
```

**Comment by brynpickering** [comment]
Aha, you probably still have an `osmox` directory in your local repo (probably empty, maybe ending in `egg_info` or similar). Ruff is getting confused by that vs the new `src/osmox` directory structure.

**Comment by brynpickering** [comment]
I've pushed an update after runnign it on my local repo

**Comment by val-ismaili** [comment]
Yeah that was it - got it working now. Thanks for bearing with me on all this. Turns out writing the tests weren't too bad, it was the other stuff i was banging my head against 😅 Also updated `test_cli.py` to use `tmp_path` so it doesn't write the files to disk.

File: tests/test_cli.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    if result.exit_code != 0:
+        traceback.print_tb(result.exc_info[-1])
+    assert result.exit_code == 0
+
+
+def test_cli_with_default_args(runner, config_path, toy_osm_path, path_output_dir):
+    # Test the command with minimal arguments
+    result = runner.invoke(cli.run, [config_path, toy_osm_path, path_output_dir])
+
+    check_exit_code(result)
```
Comment (brynpickering) [inline tests/test_cli.py:48]: Could you check that the file has actually saved?

You could do:
```python
out_file = Path(path_output_dir / filename + f".{output_format_mapped})
assert out_file.exists()
```
Would need the `filename` from the `run` function and to map `output_format` to expected file suffix.

File: tests/test_cli.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    assert result.exit_code == 0
+
+
+@pytest.mark.parametrize("output_format", ["geojson", "geopackage", "geoparquet"])
+def test_cli_output_formats(runner, config_path, toy_osm_path, path_output_dir, output_format):
+    result = runner.invoke(
+        cli.run,
+        [config_path, toy_osm_path, path_output_dir, "-f", output_format, "-crs", "epsg:4326"],
```
Comment (brynpickering) [inline tests/test_cli.py:56]: Add a test with non-epsg:4326 and check that it has both orig crs and standardised crs files saved.

**Review by brynpickering — CHANGES_REQUESTED** [review]
Nice, it's getting there!

Now that you have the hang of the tests, there's just some improvements to make to those that are there and one more I think is worth adding to match your code changes. Hopefully they won't be too difficult to implement.

**Comment by val-ismaili** [comment]
I've added a test as per your comments for testing default crs is still saved when user enters a custom crs. And then also adding to the existing test. I've put in fixtures where I thought made sense.

File: tests/test_cli.py [hunk 6]
```diff
@@ -57,3 +71,18 @@ def test_cli_output_formats(runner, config_path, toy_osm_path, path_output_dir,
...
+@pytest.mark.parametrize("crs", ["epsg:27700"])
+def test_cli_output_crs(
+    runner, config_path, toy_osm_path, path_output_dir, crs, default_output_file_path
+):
+    result = runner.invoke(cli.run, [config_path, toy_osm_path, path_output_dir, "-crs", crs])
+    check_exit_code(result)
+    assert result.exit_code == 0
+
+    # test that file with default crs is still produced
+    assert default_output_file_path.exists()
```
Comment (brynpickering) [inline tests/test_cli.py:85]: Shouldn't it be a different file path with different output formats? This is just checking for `.gpkg`

Comment (val-ismaili) [reply tests/test_cli.py:85]: So `default_output_file_ending()` fixture is `_epsg_4326.gpkg` as default file format and crs is `.gpkg` and `epsg:4326`. As this test is specifically changing crs and just using default file format the expected output will always be `_epsg_4326.gpkg`

Comment (val-ismaili) [reply tests/test_cli.py:85]: The combination of crs and output format tests means its always testing that a file with default crs will be saved and then also different file formats will be saved

Comment (brynpickering) [reply tests/test_cli.py:88]: Ah yeah, you're right. I missed the part where some code was hidden so thought you were applying it to the test where you change the file extension!

**Review by brynpickering — APPROVED** [review]
Looking good!

**Comment by brynpickering** [comment]
I added a check for file existence when changing the file extension. I've also updated the config now that you've added these tests and raised project coverage from 60% to 84% (@mfitz won't be happy that it is through an integration test rather than unit tests, but we'll let that slide ;) ).

**Pull request merged**